src/core/constraint.jl

```diff
@@ -170,22 +183,71 @@ function constraint_storage_state(pm::GenericPowerModel, n_1::Int, n_2::Int, i::
     sd_2 = var(pm, n_2, :sd, i)
     se_2 = var(pm, n_2, :se, i)
     se_1 = var(pm, n_1, :se, i)
+
     JuMP.@constraint(pm.model, se_2 - se_1 == time_elapsed*(charge_eff*sc_2 - sd_2/discharge_eff))
 end
 
 ""
 function constraint_storage_complementarity(pm::GenericPowerModel, n::Int, i)
     sc = var(pm, n, :sc, i)
     sd = var(pm, n, :sd, i)
+
     JuMP.@constraint(pm.model, sc*sd == 0.0)
 end
 
+""
+function constraint_storage_complementarity_mi(pm::GenericPowerModel, n::Int, i, charge_ub, discharge_ub)
+    sc = var(pm, n, :sc, i)
+    sd = var(pm, n, :sd, i)
+    sc_on = var(pm, n, :sc_on, i)
+    sd_on = var(pm, n, :sd_on, i)
+
+    JuMP.@constraint(pm.model, sc_on + sd_on == 1)
+    JuMP.@constraint(pm.model, sc_on*charge_ub >= sc)
+    JuMP.@constraint(pm.model, sd_on*discharge_ub >= sd)
+end
+
 ""
 function constraint_storage_loss(pm::GenericPowerModel, n::Int, i, bus, r, x, standby_loss)
     vm = var(pm, n, pm.ccnd, :vm, bus)
     ps = var(pm, n, pm.ccnd, :ps, i)
     qs = var(pm, n, pm.ccnd, :qs, i)
     sc = var(pm, n, :sc, i)
     sd = var(pm, n, :sd, i)
+
     JuMP.@NLconstraint(pm.model, ps + (sd - sc) == standby_loss + r*(ps^2 + qs^2)/vm^2)
 end
+
+""
+function constraint_storage_on_off(pm::GenericPowerModel, n::Int, i, pmin, pmax, qmin, qmax, charge_ub, discharge_ub)
+    z_storage = var(pm, n, :z_storage, i)
+    ps = var(pm, n, pm.ccnd, :ps, i)
+    qs = var(pm, n, pm.ccnd, :qs, i)
+
+    JuMP.@constraint(pm.model, ps <= z_storage*pmax)
+    JuMP.@constraint(pm.model, ps >= z_storage*pmin)
+    JuMP.@constraint(pm.model, qs <= z_storage*qmax)
+    JuMP.@constraint(pm.model, qs >= z_storage*qmin)
+end

```

Comment:
constraint_storage_on_off has been updated with variations:

AbstractPowerModel: sets ps and qs to zero when off. sc, sd are free to represent loss
AbstractActivePowerModel: only sets ps because qs does not exist
DCPlossless: sets ps, sd, sc to zero to tighten formulation.

@ccoffrin what files should these be placed in?

Reply:
The AbstractPowerModel version is the most general so that goes in base/constraint.  The AbstractActivePowerModel goes in form/apo.jl (near the current storage constraints).  The lossless version can go in form/dcp.jl